recovery/elliptics_recovery/types/merge.py

```diff
@@ -733,114 +270,154 @@ def recover(self, keys):
+            for response in responses:
+                if self._check_bad_key(response):
+                    bad_keys.append((key, ) + response)
+
+        for attempt in range(self.ctx.attempts):
+            if not bad_keys:
+                break
+
+            results = []
+            for key, _, addr, backend_id in bad_keys:
+                self.remove_session.set_direct_id(addr, backend_id)
+                result = self.remove_session.remove(elliptics.Id(key))
+                results.append(result)
+
+            timeouted_keys = []
+            is_last_attempt = (attempt == self.ctx.attempts - 1)
+            for i, r in enumerate(results):
+                status = r.get()[0].status
+                log.info("Removing key: {0}, status: {1}, last attempt: {2}".format(bad_keys[i], status, is_last_attempt))
+                if status == -errno.ETIMEDOUT:
+                    timeouted_keys.append(bad_keys[i])
+            bad_keys = timeouted_keys
 
     def _check_bad_key(self, response):
-        status = response[1]
-        return status == -errno.EBADFD or status == -errno.EILSEQ
+        status = response[0]
+        return status in (-errno.EBADFD, -errno.EILSEQ)
 
-    def _get_unrecovered_keys(self, responses):
+    def _has_unrecovered_keys(self, responses):
         '''
-        Returns keys that was not recovered via server-send.
+        Returns True, if some key was not recovered via server-send.
         '''
-        keys = []
-        for val in responses.iteritems():
-            key_responses = val[1]
-            if not key_responses or self._check_unrecovered_key(key_responses):
-                key = elliptics.Id(val[0])
-                keys.append(key)
-        return keys
+        for key_responses in responses.itervalues():
+            if self._check_unrecovered_key(key_responses):
+                return True
+        return False
 
     def _check_unrecovered_key(self, responses):
         '''
-        Returns True, if a valid key exists on the backend, but the key could not be recovered by any reason.
+        Returns True, if a valid key exists at the backend, but the key could not be recovered by any reason.
         '''
         for r in responses:
-            status = r[1]
+            status = r[0]
             if status < 0 and status != -errno.ENOENT and not self._check_bad_key(r):
                 return True
         return False
 
+    def _update_stats(self, start_time, processed_keys, recovers_in_progress, status):
+        speed = processed_keys / (time.time() - start_time)
+        recovers_in_progress -= processed_keys
+        self.stats.set_counter('recovery_speed', round(speed, 2))
+        self.stats.set_counter('recovers_in_progress', recovers_in_progress)
+        if status != -errno.ETIMEDOUT:
+            self.stats.counter('recovered_keys', 1 if status == 0 else -1)
+            self.ctx.stats.counter('recovered_keys', 1 if status == 0 else -1)
```

Comment:
I saw these fields in #668 pull request, why do you update them in two different places? Is this the same code as in #668 ?